scalyr_agent/agent_main.py

```diff
@@ -1069,7 +1069,527 @@ def __get_log_files_initial_positions(self, only_new_files=False):
+                backup_count=self.__config.log_rotation_backup_count,
+                logs_directory=self.__config.agent_log_path,
+                agent_log_file_path=AGENT_LOG_FILENAME,
+            )
+
+            self.__update_debug_log_level(
+                self.__config.debug_level, self.__config.debug_level_logger_names
+            )
+
+            # We record where the log files(including multiprocess worker sessions logs) currently are
+            # so that we can (in the worse case) start copying them from those position.
+            logs_initial_positions = self.__get_log_files_initial_positions()
+
+            start_up_msg = scalyr_util.get_agent_start_up_message()
+            log.info(start_up_msg)
+            log.log(scalyr_logging.DEBUG_LEVEL_1, start_up_msg)
+
+            # Log warn message if non UTF-8 locale is used - this would cause issues when trying
+            # to monitor files with unicode characters inside the file names or inside the
+            # content
+            _, encoding, _ = scalyr_util.get_language_code_coding_and_locale()
+            if encoding.lower() not in ["utf-8", "utf8"]:
+                if sys.platform.startswith("win"):
+                    log.warn(NON_UTF8_LOCALE_WARNING_WINDOWS_MESSAGE % (encoding))
+                else:
+                    log.warn(NON_UTF8_LOCALE_WARNING_LINUX_MESSAGE % (encoding))
+
+            self.__controller.emit_init_log(log, self.__config.debug_init)
+
+            self.__start_or_stop_unsafe_debugging()
+
+            raw_scalyr_server = self.__config.raw_scalyr_server
+            self.__print_force_https_message(
+                self.__config.scalyr_server, raw_scalyr_server
+            )
+
+            last_config_change_check_time = (
+                last_essential_monitors_check_time
+            ) = time.time()
+
+            # Create generator that has to check config for changes and, if so, reload everything which is needed
+            # according to a new configuration.
+            check_config_for_change_gen = self._check_config_change(
+                logs_initial_positions=logs_initial_positions
+            )
+
+            # Advance generator to the beginning of the first loop iteration.
+            next(check_config_for_change_gen)
+
+            try:
+                while not self.__run_state.sleep_but_awaken_if_stopped(0.1):
+
+                    # Check config for changes when it's time.
+                    if (
+                        time.time()
+                        >= last_config_change_check_time
+                        + self.__config.config_change_check_interval
+                    ):
+                        next(check_config_for_change_gen)
+                        last_config_change_check_time = time.time()
+
+                    # Check for essential monitors every 5 sec.
+                    if time.time() >= last_essential_monitors_check_time + 5:
```

Comment:
Is there a chance that the code above could take 5 seconds or longer and we would always end up checking those essential monitors and code busy waiting / 100% cpu usage?

I don't think so, but wanted to double check.

Reply:
It seems that it shouldn't, since we call new `time.time()` on each check, so we don't give any chance for the usage of the outdated timestamp.